### ISSUE TITLE
addr2line -f fails with "InvalidDebugSymbols(SubroutineMissingName(0, 100459))"

### ISSUE THREAD

**Issue by slonopotamus** [gimli-rs/addr2line#35]
Way to reproduce:

    wget http://slonopotamus.org/libUE4Editor-Core.so
    addr2line -f -e libUE4Editor-Core.so 0x2c09b2

Expected output (as produced by `addr2line` from GNU binutils and gdb, and I know this is correct result):

    _Z20PlatformCrashHandleriP9siginfo_tPv
    /home/marat/production/ue4/Engine/Source/Runtime/Core/Private/Linux/LinuxPlatformCrashContext.cpp:595

Actual output:

    addr2line: Error(Msg("failed to analyze debug information"), State { next_error: Some(Error(Msg("encountered invalid compilation unit"), State { next_error: Some(Error(InvalidDebugSymbols(SubroutineMissingName(0, 100459)), State { next_error: None, backtrace: None })), backtrace: None })), backtrace: None })

Note that running without `-f` switch properly resolves filename/lineno, it's only function name that fails to resolve.

Funny facts:

1. `addr2line` from `elfutils` [works in the opposite way][1]: it resolves function name but fails with filename/lineno.
2. `addr2line` from `elftoolchain` [segfaults][2] when trying to resolve function name, and [gives wrong filename/lineno][3].
3. `addr2line` from FreeBSD [gives wrong filename/lineno and fails to find function name][4]

So, among five `addr2line` implementations, there are no two that behave the same way.

[1]: https://sourceware.org/bugzilla/show_bug.cgi?id=21247
[2]: https://sourceforge.net/p/elftoolchain/tickets/546/
[3]: https://sourceforge.net/p/elftoolchain/tickets/545/
[4]: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=217736

**Comment by philipc** [comment]
I think we need to follow `DW_AT_specification` references, similar to the way we currently handle `DW_AT_abstract_origin`.

**Comment by slonopotamus** [comment]
I've tested the fix you provided in #36 and indeed `addr2line` now can resolve function name, however it takes pretty lots of time to do that.

binutils:

    $ time addr2line -f -e libUE4Editor-Core.so 0x2c09b2
    _Z20PlatformCrashHandleriP9siginfo_tPv
    /home/marat/production/ue4/Engine/Source/Runtime/Core/Private/Linux/LinuxPlatformCrashContext.cpp:595
    
    real    0m0.085s
    user    0m0.060s
    sys     0m0.020s

gimli-rs (release):

    $ time ./target/release/addr2line -f -e libUE4Editor-Core.so 0x2c09b2
_Z20PlatformCrashHandleriP9siginfo_tPv
    /home/marat/production/ue4/Engine/Source/Runtime/Core/Private/Linux/LinuxPlatformCrashContext.cpp:595

    real    0m0.321s
    user    0m0.310s
    sys     0m0.010s

gimli-rs (debug):

    $ time ./target/debug/addr2line -f -e libUE4Editor-Core.so 0x2c09b2
    _Z20PlatformCrashHandleriP9siginfo_tPv
    /home/marat/production/ue4/Engine/Source/Runtime/Core/Private/Linux/LinuxPlatformCrashContext.cpp:595

    real    0m9.073s
    user    0m8.830s
    sys     0m0.020s

**Comment by philipc** [comment]
debug builds will be slow. We don't care about performance for it.

The release build is slower for a single address, but faster for multiple addresses. Here's the benchmark for the same library testing 26069 different addresses:

````
==> Benchmarking
binutils  nofunc  4.00  172868
master    nofunc  0.22  30228
==> Benchmarking with -f
binutils  func  4.10   172852
master    func  1.19   70344
````

Is performance for single addresses something you care about? There's definitely room for improvement there, but not sure if it is worth the effort. If you do care about it, please create a separate issue.

**Comment by philipc** [comment]
Oh, for that benchmark, the third column is time in seconds, the fourth column is memory usage in bytes.

**Comment by slonopotamus** [comment]
For future readers: actually, there is one more `addr2line` implementation (although with different command line args) - `llvm-symbolizer`. It is consistent with binutils and (after #36) also with gimli.
